File: GameLibrary/src/cc/lib/swing/AWTToggleButton.java
package cc.lib.swing;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.JCheckBox;

import cc.lib.ui.IButton;

public class AWTToggleButton extends JCheckBox implements ActionListener {

    private boolean ignore = false;

    public AWTToggleButton(String text) {
        this(text, false);
    }

    public AWTToggleButton(IButton button) {
        this(button.getLabel(), false);
        setToolTipText(button.getTooltipText());
    }

    public AWTToggleButton(String text, boolean selected) {
        super(text);
        setSelected(selected);
        addActionListener(this);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        if (!ignore) {
            updateUI();
            onToggle(isSelected());
        }
    }

    @Override
    public void setSelected(boolean selected) {
        ignore = true;
        super.setSelected(selected);
        ignore = false;
    }

    protected void onToggle(boolean on) {}
}

File: SwingApps/src/cc/lib/swing/AWTToggleButton.kt
package cc.lib.swing

import cc.lib.ui.IButton
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.JCheckBox

open class AWTToggleButton @JvmOverloads constructor(text: String?, selected: Boolean = false) : JCheckBox(text), ActionListener {
	private var ignore = false

	constructor(button: IButton) : this(button.getLabel(), false) {
		toolTipText = button.getTooltipText()
	}

	init {
		isSelected = selected
		addActionListener(this)
	}

	override fun actionPerformed(e: ActionEvent) {
		if (!ignore) {
			updateUI()
			onToggle(isSelected)
		}
	}

	override fun setSelected(selected: Boolean) {
		ignore = true
		super.setSelected(selected)
		ignore = false
	}

	protected open fun onToggle(on: Boolean) {}
}